docs/src/main/asciidoc/mongodb-panache.adoc

```diff
@@ -378,6 +381,9 @@ personRepository.delete("status", Status.Alive);
 
 // delete all persons
 personRepository.deleteAll();
+
+// update all living persons
```

Comment:
This looks like you mixed up the query/update part, no? It's not updating all living persons, but all persons with a `Moral` name (does that even compile?) to make them `Alive`, no?

That kinda proves the API is confusing, no? It's not clear to me if the query or update should be first.

Reply:
Query is first, then the update document (query is a filter in MongoDB terms), what is confusing is that you can pass parameters on both side and it takes the parameters as if the query is a unique one ... Using varags parameters it is not very readable.

Maybe we need to introduce an `UpdateQuery` object that will allow to intoduce the two "query part" one by one
```
Person.update("name = ?1", "Mortal").with("status = ?1", Status.Alive);
```

WDYT ?

Reply:
I like that idea. That's more readable. It also supports shortcut syntax: `Person.update("name", "Mortal").with("status", Status.Alive);`

Reply:
OK, I'll give it a try

Reply:
The new impl is done, can you have a look ?

Reply:
The comment here (and above) are still wrong: this does not update all living persons. It sets all persons named "Mortal" to living.

Reply:
@emmanuelbernard can you give this an eye? I think it looks good, but it's different to the ORM version. Given that queries can be native here, I think it's fine.

Reply:
> The comment here (and above) are still wrong: this does not update all living persons. It sets all persons named "Mortal" to living.

This was copied from the Hibernate with Panache documentation, I'll change it there also.

Reply:
Hah, I was afraid this would turn out to be the case ;) Thanks!

Reply:
Are you asking me for my opinion on

`update("name", "Mortal").with("status", Status.Alive);`
being different than
`update("name = 'Mortal' where status = ?1", Status.Alive);`

Would be cool to be unified but I imagine there is a reason behind that difference?

What makes me uneasy is how can I do a update all name to "Mortal" without condition of being alive? And the counter reason is how is it clear that I am not changing all the DB by accident by forgetting the "where clause" - that could not have happened ever to one of us, right so this is purely hypothetical that an ecomerce site did update its catalog with price=0 for all products ;P

Reply:
@emmanuelbernard in fact the implementation is
 `update(filter, params).with(document, params)`

So 
`update("name", "Mortal").with("status", Status.Alive);`

Correspond to the following one implemented on Hibernate with Panache (but cannot be done easily with MongoDB)
`update("name = ?1 where status = ?2", "Mortal", Status.Alive);`

To update without filter you could use the empty String or null (worth testing it)
`update("").with("status", Status.Alive);`

We can also add an overload of update with no parameters for this case, good point :)

Reply:
> Would be cool to be unified but I imagine there is a reason behind that difference?

Yeah, and the reason is that both the query and the update need to support native queries so we can't have them as a single `String` param, and having two `String` params for the query and update parts is confusing (their order is).

Now, you're right that we do need another version with no query part. We can't go with `update("").with(...)` because it looks silly. We could do `update().with(...)` but it pains me that we're going through all that trouble for native queries, which in theory is the less common case.

At this point, we could have `update().with(update)` and `update(query).with(update)` OR `update(panacheQlQuery)` and `updateNative(nativeQuery?).with(updateQuery)`.

Not sure yet which I prefer.

Reply:
We choose to swith panacheql/nativeql by inspecting the query string for the presence of `{`. 
I would really prefere to keep it this way so I favor `update().with(update, params)`.

The issue is that a MongoDB update is two queries and not one (a filter query and an update document), and we need to provide parameters substitution for both.

As an alternative (that I don't really support), we can add `set` support for the current panacheQL so we can write a MongoDB update query as `update("query with set", params)` method. But I feel this would not be natural for MongoDB users. 

In this case, native query could be supported by providing the , inside the first varags parameters, the update document (a String or a BSON Document) ...

Examples:
```
update("where name=?1 set status=?2", "Mortal", Status.ALIVE);
update("{'name'=?1}", "{'status'=?2"}, "Mortal", Status.ALIVE);
```

For this example to works we need to support `set` inside the PanacheQL parser and the MongoParserVisitor.

As I said, I'm not a big fan of it and prefere to stick with `update(query, param).with(document, param)`.

Reply:
> @emmanuelbernard in fact the implementation is
> `update(filter, params).with(document, params)`
> 
> So
> `update("name", "Mortal").with("status", Status.Alive);`
> 
> Correspond to the following one implemented on Hibernate with Panache (but cannot be done easily with MongoDB)
> `update("name = ?1 where status = ?2", "Mortal", Status.Alive);`
> 
> To update without filter you could use the empty String or null (worth testing it)
> `update("").with("status", Status.Alive);`
> 
> We can also add an overload of update with no parameters for this case, good point :)

I am a bit confused. What you call filter is not the filter but the transformation clause. And the narrowing (filtering) is in the with operator? The way you describe it seems to be the opposite but your examples seems ot disagree with your text. In my head that is.

Reply:
So I think it's likely ok to have the double methods currently explored. But you guys still confuse the hell out of me and I still can't figure out where to put my transformations and where to put my narrowing predicate. Which makes me feel the names can be improved. Proposed alternative names

- update(...).where(...) (e.g. update("some field changes").where("some restrictions on fields") with where being optional
- update(...).for(...) with same semantic above
-

Reply:
**This is** `update(filter, params).with(newDocument, params)`. The fact that you are confuse show that the examples was not clear and maybe the API is not clear enougth !

Keep in mind that a typical MongoDB multi-update looks like this (from the mongo shell or a tool like Robot3T) :
```
dp.persons.update(
    {"name": "Mortal"}, 
    {"status": "Alive"},
    {"multi": true}
); 
```
This is a JavaScript method call with 3 parameters: filter, update document and options.

We can replace it with `update().filter(filter, params).withDocument(newDocument, params)` so that it is be more clear and the filter could be optional to provides update on all documents.

With this new API, there will be one more method call ...

The other advantage of this new API is that we can also add it to Hibernate with Panache if some people think it will be a good addition to the existing multi-update support :)

Reply:
So I guess you copied the notion of filter and document from Mongo (where document is roughly equivalent to SQL SET section - again if I understood but I’m still not sure). So the core or my problem is that the Mongo API confuses me. I would prefer the semantic I proposed earlier (update and where methods) which align with SQL and JP-QL and feel more natural to me.

Reply:
Yes, your guess is right :)

I'm OK with `update(updateDocument, params).where(filter, params);`, it will be a little less natural for peoples used to MongoDB but is still easy to understand.

For the records Spring Data do it this way:
```
mongoTemplate.updateMulti(
    new Query(where("accounts.accountType").is(Account.Type.SAVINGS)),
    new Update().inc("accounts.$.balance", 50.00), 
   Account.class
);
```

Reply:
Can you do the same query but with our API, with parameters and all? Looks like it will be more compact and readable.

Reply:
The Spring data example is based on their QueryDsl, a type safe query DSL that is very verbose.

I think that our aproach of update().with() or update().where() better match the Panache way (compact, defines query with strings, parameters, ...).

So I can propose the following API:
```
update(filter, params).with(updateDocument, params)// 1
update(updateDocument, params).where(filter, params)// 2
update().filter(filter, params).set(updateDocument, params)// 3
```

We can change the name of the method (I use `filter` because it matches MongoDB terms but you can use `where` instead).
If we plan to integrate the same facility inside Hibernate with Panache in addition to the current multi-update we need to choose terms that are not tied to SQL nor MongoDB ...

Reply:
2

Reply:
So you don't want a panache-ql version after all?

How about when there's no filter?

One possibility would be:

```java
update("foo = ?1 where foo is null", something) // panache-ql
updateNative("{'foo'=?1}", something).where("{'status'=?1"}, Status.Alive); // native update queries
updateNative("{'foo'=?1}", something).all(); // or some other no-filter term
```

Reply:
`update("foo = ?1 where foo is null", something)`
This didn't feels right to me, it's realy not the way to do it in MongoDB. And we cannot have a panache-ql version and the `update(update, params).where(query, params)` that we discuss with  @emmanuelbernard and the last one has my preference.

I would also prefere to keep the same methods for both native and non-native query as we did for all the other methods, at least inside `PanacheMongoEntityBase` and `PanacheMongoRepositoryBase` that already contains a lot of methods.

For the case there is no filter, I add a `update(update, params).where()` I ask in a comment wheter to have an empty `where()` or to have `all()` instead, none have my preference so you can choose ...